### PR TITLE
Test and fix QgsRectangle::isNull

### PR DESCRIPTION
**Pull request by strk**
Adds tests for QgsRectangle::isNull, improve toString to show the difference between Null and Empty, eventually fix the 0,0,0, rectangle case to still be Empty and NOT Null (to match documentation).

Superceeds #45607

Aims at closing #45563

### CODE REVIEW

File: src/core/geometry/qgsrectangle.h [hunk 1]
```diff
@@ -479,8 +484,7 @@ class CORE_EXPORT QgsRectangle
     bool isNull() const
     {
       // rectangle created QgsRectangle() or with rect.setMinimal() ?
-      return ( qgsDoubleNear( mXmin, 0.0 ) && qgsDoubleNear( mXmax, 0.0 ) && qgsDoubleNear( mYmin, 0.0 ) && qgsDoubleNear( mYmax, 0.0 ) ) ||
```
[the remark this answers is not in the document]
Comment (strk) [reply src/core/geometry/qgsrectangle.h:482]: The goal of this PR is to have implementation match documentation, see https://github.com/qgis/QGIS/issues/45563 -- if plugin/script authors coded following documentation I'd think they should accept this.

**Comment by strk** [comment]
I've rebased this PR against current master branch. It's a pity that the stale bot kicked in beacause we're in feature freeze again

**Comment by strk** [comment]
Still useful

**Comment by Gustry** [comment]
> It's a pity that the stale bot kicked in beacause we're in feature freeze again

The feature freeze is not a blocker for continuing the development on the branch.
It's just a blocker to merge a PR in a frozen branch.

If the PR is ready to be merged (approved, reviewed etc), then the label "merge after thaw" can be applied. The bot won't close PR with this labels.

My 2 cents ;-)

**Comment by nyalldawson** [comment]
Come on @strk, by what definition of stale ISN'T this a stale pr? It hasn't been updated since February and is no closer to being merge ready since then. A ~4 month old pr which is not being actively contributed to is 100% stale. 

Let it close, and reopen when you're ready to move this forward.

**Comment by strk** [comment]
On Wed, Jun 22, 2022 at 02:10:06PM -0700, Nyall Dawson wrote:
> Come on @strk, by what definition of stale ISN'T this a stale pr? It hasn't been updated since February and is no closer to being merge ready since then. A ~4 month old pr which is not being actively contributed to is 100% stale. 

I'm not debating the definition of Stale. This is indeed Stale, but
why closing it ? Can't you just put Stale PRs out of your sight
by filtering them out with the GitHub filter ?

> Let it close, and reopen when you're ready to move this forward.

Whatever, it could be NEVER. Fact is: QGIS API documentation is
currently NOT matching the implementation, either the documentation
is broken or the implementation is. Closing this out won't really 
help in fixing this problem.

--strk;

**Comment by nyalldawson** [comment]
> Can't you just put Stale PRs out of your sight
by filtering them out with the GitHub filter ?

And on the contrary, can't YOU just adapt to the processes and policies which the most active contributors to this project have all agreed to? Why must you make this such an uphill battle? 

You're a valuable contributor to this project, but boy, you take up a LOT of emotional effort -- and it just doesn't need to be that way at all. 🤷

**Comment by 3nids** [comment]
Let's see at least if this PR passes the tests 👓

**Comment by strk** [comment]
Rebased, continues in #53323

**Pull request closed without merging**